Core/include/Acts/Vertexing/IVertexFinder.hpp

```diff
@@ -0,0 +1,40 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Utilities/Any.hpp"
+#include "Acts/Utilities/Result.hpp"
+#include "Acts/Vertexing/VertexingOptions.hpp"
+
+#include <vector>
+
+namespace Acts {
+
+class Vertex;
+struct InputTrack;
+struct VertexingOptions;
+
+class IVertexFinder {
+ public:
+  using State = Acts::AnyBase<128>;
+
+  virtual Result<std::vector<Vertex>> find(
+      const std::vector<InputTrack>& trackVector,
+      const VertexingOptions& vertexingOptions, State& state) const = 0;
+
+  virtual State makeState() const = 0;
+
+  virtual bool hasTrivialState() const = 0;
+
+  virtual void setTracksToRemove(
+      State& state, const std::vector<InputTrack>& removedTracks) const = 0;
```

Comment:
why "set" and not `removeTracks`?

Reply:
Because the way the removal works currently is that the e.g. `AdaptiveGridDensityVertexFinder` removes these tracks in the next find call. We could think about collapsing this to doing the track removal when the function is called.

Reply:
So, looking at this more, it turns out that the `GridDensityVertexFinder` uses the result of the track removal (i.e. "no more tracks were removed") to signal a break to the parent finder. `AdaptiveGridDensityVertexFinder` handles this differently, though. I need to think more about this.

Reply:
I'd prefer to leave this for this PR, and then rework how the grid density finders handle track removal and termination.